README
Gentoo::PerlMod::Version

This module acts as a reference implementation of how Gentoo maps CPAN
and Perl versions, and transforms them into derived versions that are
suitable for Gentoo dependency tracking.

Perl has several primary formats of versions, the most notable one being
"float" style versions, in the form "x.yyyyyyyyyyy" where the number of
"y"'s are arbitrary, and are interpreted as a floating point value.

That is, 1.001 is NOT the same as 1.01 and 1.1

However, Gentoo's version scheme sees 1.001 similar to 1.001.000 which
is similar to 1.1.0 and thus, similar to 1.1.

Obviously this will not do, because when somebody says they need ">=1.05
(g:1.5)" expecting "1.06 (g:1.6)", but instead get "1.009 (g:1.9)",
things will break.

Hence, detection of these cases and normalizing them is essential:

  1.001 -> 1.1.0
  1.01  -> 1.10.0
  1.1   -> 1.100.0
  1.05  -> 1.50.0
  1.06  -> 1.60.0
  1.009 -> 1.9.0

  1.9.0 < 1.50.0 < 1.60.0

The simplest use of this library is with the shipped tool,
"gentoo-perlmod-version.pl"

  gentoo-perlmod-version.pl --oneshot 1.06  # 1.6.0

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for Gentoo-PerlMod-Version

v0.8.2 2017-03-05T11:52:15Z b2dce30

v0.8.1 2017-03-05T11:11:43Z 2465f66
 [Bugfix]
 - Fix test problems with -Ddefault_inc_excludes_dot
 - Closes https://github.com/kentnl/Gentoo-PerlMod-Version/issues/1

 [Dependencies::Stats]
 - Dependencies changed since 0.8.0, see misc/*.deps* for details
 - develop: +8 ↑2 (suggests: ↑2)

 [Misc]
 - Version scheme now includes a "v" due to upstream tooling constraints

0.8.0 2015-07-16T23:24:29Z 301fe54
 [00 Major / Features]
 - gentoo-perlmod-version.pl is now fatpacked.
 - This means the code should run better in mixed environments as well as between perl upgrades

 [Dependencies::Stats]
 - Dependencies changed since 0.7.1, see misc/*.deps* for details
 - develop: +2 -1 (suggests: ↑1)

0.7.1 2015-06-23T03:30:53Z 8b35638
 [00 Minor]
 - Avoid using an XS dep to simply iterate an array pairwise.
 - This effectively halves our total runtime dependencies.

 [Dependencies::Stats]
 - Dependencies changed since 0.7.0, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +10 ↑3 -3 (recommends: +1, suggests: ↑2)
 - runtime: -1
 - test: +1 (recommends: ↑1 ↓1)

0.7.0 2014-09-27T07:40:21Z
 [00 Important - Installing Toolkit Change]
 - This may affect hand operated downstreams
 - Code is migrated from using Module::Build to using ExtUtils::MakeMaker as an install tool.
 - https://questhub.io/realm/perl/quest/5388cc12bbd0be5b2500011f

 [01 Migrated to Sub::Exporter::Progressive]
 - No fancy stuff is really required internally, so a lighter weight Sub::Exporter::Progressive is now in use.

 [Dependencies::Stats]
 - Dependencies changed since 0.6.0, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +50 -1 (recommends: -1, suggests: +1 ↑1)
 - runtime: +2 ↓1 -1
 - test: +2 ↓1 -2 (recommends: +3)

 [Internals]
 - Fatal Exceptions don't lose their context as soon as they're thrown.
 - However, backtraces are subsequently slightly impeded.
 - Use Devel::Cover for full backtraces.

 [Metadata]
 - recognize 1x contributor
 - migrate to kentnl org

0.6.0 2012-10-27T12:27:17Z
 [00 IMPORTANT CHANGES]
 - $perlver being undef now is an error.
 - All error code is now refactored into the ::Error class, which ultimately grants more flexibility and better code
   organisation.
 - %ENV interaction is now handled vis a few utility functions in the ::Env class.
 - Functions in ::Env and ::Error are proxied into the ::Version class via some lazy-load functions.

 [Dependencies::Stats]
 - Dependencies changed since 0.5.0, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: (suggests: ↑1)
 - test: +1 ↑1

 [Documentation]
 - Miscelaneous improvements, s/method/func/, and more general coverage of some error functions.

0.5.0 2012-06-08T12:08:07Z
 [00 IMPORTANT CHANGES]
 - Errors that occur when a version fails to validate have changed in some circumstances.
 - Now reports the version token that failed to validate.
 - laxness can be overridden globally via new GENTOO_PERLMOD_VERSION_OPTS=" always_lax=$n "
 - user provided data can now be stripped from error output by GENTOO_PERLMOD_VERSION_OPTS=" taint_safe "
 - carp messages can be more verbose about context/options by setting GENTOO_PERLMOD_VERSION_OPTS=" carp_debug "

 [Dependencies::Stats]
 - Dependencies changed since 0.4.0, see misc/*.deps* for details
 - develop: (recommends: ↑1, suggests: ↑1)

 [Internals]
 - Carp now loaded on-demand.
 - List::MoreUtils now loaded on demand, which is good, because its only used at lax = 2
 - croak is proxied via _fatal() which takes lots more information.

0.4.0 2012-03-01T08:37:02Z
 [00 IMPORTANT CHANGES]
 - new 'oneshot' mode for validating/normalising a single version, for use predominantly in other scripts like bash.
 - usage: VALUE="$( gentoo_perlmod_version.pl --oneshot "$VERSION" )"

 [Dependencies::Stats]
 - Dependencies changed since 0.3.1, see misc/*.deps* for details
 - develop: (suggests: ↑1)
 - runtime: +1

 [Documentation]
 - Fixed a few typos and clarified some minor things.

 [Packaging]
 - Update LICENSE ( Year, Indentation )

 [Tests::New]
 - minimum-version.t

0.3.1 2011-10-22T14:35:08Z
 - Maintenance Release

 [Dependencies::Stats]
 - Dependencies changed since 0.3.0, see misc/*.deps* for details
 - runtime: +2
 - test: -1

 [Internals]
 - VERSION no longer declared in BEGIN
 - All Packages declare $AUTHORITY

 [Packaging]
 - Move Changelog to CPAN::Changes format.
 - Move author/release tests to xt/
 - Update LICENSE ( Indentation, Address )
 - Git Repo URI's are now all https://
 - Added x_authority

0.3.0 2011-02-26T02:08:13Z
 - Switch to @Author::KENTNL bundle.
 - Don't strip excess 0 groups.
 - Use standard 'normalized' dotted-decimal versions ( without leading 'v' and with at least 3 components ) ( Thanks to
   Torsten Veller )

 [Dependencies::Stats]
 - Dependencies changed since 0.2.2, see misc/*.deps* for details
 - develop: +1 -1 (recommends: +1 -1, suggests: +1 -1)

0.2.2 2011-01-11T14:20:34Z
 - Added POD to the script to be helpful.
 - Removed a few errors from the POD.
 - Hack it so lax version doesn't influence the whole parse, instead, laxing is only applied when its deemed the input
   isn't within a defined format.
 - Dropped _has_bad_bits in favour of individual targeted regex.

0.2.1 2011-01-10T21:14:42Z
 - Fix a minor issue where v-strings were rejected because we thought they were ascii. Thanks to Torsten Veller for
   reporting.

0.2.0 2011-01-09T22:26:45Z
 - Managed to accidentally sneak the last one out without the dep on version.pm.
 - Redid the whole version split->lint-tidyup chain w/ Vincent Pits code. ( Thanks VPIT )

 [Dependencies::Stats]
 - Dependencies changed since 0.001.000, see misc/*.deps* for details
 - runtime: +1

0.001.000 2011-01-09T19:24:56Z
 - First version, released on an unsuspecting world.


lib/Gentoo/PerlMod/Version.pm
use 5.006;
use strict;
use warnings;

package Gentoo::PerlMod::Version;

our $VERSION = 'v0.8.2';

# ABSTRACT: Convert arbitrary Perl Modules' versions into normalized Gentoo versions.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Sub::Exporter::Progressive -setup => { exports => [qw( gentooize_version )] };
use version 0.77;

sub gentooize_version {
  my ( $perlver, $config ) = @_;
  $config ||= {};
  if ( not defined $perlver ) {
    return _err_perlver_undefined($config);
  }
  $config->{lax} = 0 unless defined $config->{lax};
  if ( _env_hasopt('always_lax') ) {
    $config->{lax} = _env_getopt('always_lax');
  }

  if ( $perlver =~ /\Av?[\d.]+\z/msx ) {
    return _lax_cleaning_0($perlver);
  }

  if ( $perlver =~ /\Av?[\d._]+(-TRIAL)?\z/msx ) {
    if ( $config->{lax} > 0 ) {
      return _lax_cleaning_1($perlver);
    }
    return _err_matches_trial_regex_nonlax( $perlver, $config );
  }

  if ( 2 == $config->{lax} ) {
    return _lax_cleaning_2($perlver);
  }
  return _err_not_decimal_or_trial( $perlver, $config );
}

###
#
# character to code translation
#
###

## no critic ( ProhibitMagicNumbers )
my $char_map = {
  ( map { $_ => $_ } 0 .. 9 ),    # 0..9
  ( map { chr( $_ + 65 ) => $_ + 10 } 0 .. 25 ),    # A-Z
  ( map { chr( $_ + 97 ) => $_ + 10 } 0 .. 25 ),    # a-z
};

# _code_for('z') -> $number
#

sub _code_for {
  my $char = shift;
  if ( not exists $char_map->{$char} ) {
    my $char_ord = ord $char;
    return _err_bad_char( $char, $char_ord );
  }
  return $char_map->{$char};
}

###
#
# Pair to number transformation.
#
#   _enc_pair( 'x','y' ) ->  $number
#
##

sub _enc_pair {
  my (@tokens) = @_;
  if ( not @tokens ) {
    return q{};
  }
  if ( @tokens < 2 ) {
    return _code_for( shift @tokens );
  }
  return ( _code_for( $tokens[0] ) * 36 ) + ( _code_for( $tokens[1] ) );
}

###
#
#  String to dotted-decimal conversion
#
# $dotstring = _ascii_to_int("HELLOWORLD");
#
###
sub _ascii_to_int {
  my $string = shift;
  my @chars = split //msx, $string;
  my @output;

  while (@chars) {
    push @output, _enc_pair( splice @chars, 0, 2, () );
  }

  return join q{.}, @output;
}

#
# Handler for gentooize_version( ... { lax => 0 } )
#
sub _lax_cleaning_0 {
  my $version = shift;
  return _expand_numeric($version);
}

#
# Handler for gentooize_version( ... { lax => 1 } )
#

sub _lax_cleaning_1 {
  my $version       = shift;
  my $isdev         = 0;
  my $prereleasever = undef;

  if ( $version =~ s/-TRIAL\z//msx ) {
    $isdev = 1;
  }
  if ( $version =~ s/_(.*)\z/$1/msx ) {
    $prereleasever = "$1";
    $isdev         = 1;
    if ( $prereleasever =~ /_/msx ) {
      return _err_lax_multi_underscore($version);
    }
  }
  $version = _expand_numeric($version);
  if ($isdev) {
    $version .= '_rc';
  }
  return $version;
}

#
# Handler for gentooize_version( ... { lax => 2 } )
#

sub _lax_cleaning_2 {
  my $version = shift;
  my $istrial = 0;

  my $has_v = 0;

  if ( $version =~ s/-TRIAL\z//msx ) {
    $istrial = 1;
  }
  if ( $version =~ s/\Av//msx ) {
    $has_v = 1;
  }

  my @parts = split /([._])/msx, $version;
  my @out;
  for (@parts) {
    if (/\A[_.]\z/msx) {
      push @out, $_;
      next;
    }
    if (/\A\d\z/msx) {
      push @out, $_;
      next;
    }
    push @out, _ascii_to_int($_);
  }

  my $version_out = join q{}, @out;
  if ($istrial) {
    $version_out .= '-TRIAL';
  }
  if ($has_v) {
    $version_out = 'v' . $version_out;
  }
  return _lax_cleaning_1($version_out);
}

#
# Expands dotted decimal to a float, and then chunks the float.
#
# my $clean = _expand_numeric( $dirty );
#
sub _expand_numeric {
  my $perlver = shift;

  my $ver = version->parse($perlver)->normal;

  $ver =~ s/\Av//msx;    # strip leading v

  my @tokens = split /[.]/msx, $ver;
  my @out;

  for (@tokens) {
    s/\A0+([1-9])/$1/msx;    # strip leading 0's
    push @out, $_;
  }

  return join q{.}, @out;
}

BEGIN {
  for my $err (qw( perlver_undefined matches_trial_regex_nonlax not_decimal_or_trial bad_char lax_multi_underscore )) {
    my $code = sub {
      require Gentoo::PerlMod::Version::Error;
      my $sub = Gentoo::PerlMod::Version::Error->can($err);
      goto $sub;
    };
    ## no critic ( ProhibitNoStrict )
    no strict 'refs';
    *{ __PACKAGE__ . '::_err_' . $err } = $code;
  }
  for my $env (qw( opts hasopt getopt )) {
    my $code = sub {
      require Gentoo::PerlMod::Version::Env;
      my $sub = Gentoo::PerlMod::Version::Env->can($env);
      goto $sub;
    };
    ## no critic ( ProhibitNoStrict )

    no strict 'refs';
    *{ __PACKAGE__ . '::_env_' . $env } = $code;
  }

}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Gentoo::PerlMod::Version - Convert arbitrary Perl Modules' versions into normalized Gentoo versions.

=head1 VERSION

version v0.8.2

=head1 SYNOPSIS

    use Gentoo::PerlMod::Version qw( :all );

    # http://search.cpan.org/~gmpassos/XML-Smart-1.6.9/
    say gentooize_version( '1.6.9' )  # 1.6.9

    http://search.cpan.org/~pip/Math-BaseCnv-1.6.A6FGHKE/

    say gentooize_version('1.6.A6FGHKE')   #  <-- death, this is awful

    # -- Work-In-Progress Features --

    say gentooize_version('1.6.A6FGHKE',{ lax => 1}) # <-- still death

    say gentooize_version('1.6.A6FGHKE',{ lax => 2}) # 1.6.366.556.632.14  # <-- the best we can do.

    say gentooize_version('1.9902-TRIAL')   #  <-- death, this is not so bad, but not a valid gentoo/stable version

    say gentooize_version('1.9902-TRIAL', { lax => 1 })   #  1.990.200_rc # <-- -TRIAL gets nuked, 'rc' is added.

=head1 DESCRIPTION

This module acts as a reference implementation of how Gentoo maps CPAN and Perl versions, and transforms
them into derived versions that are suitable for Gentoo dependency tracking.

Perl has several primary formats of versions, the most notable one being C<float> style versions, in the form
C<x.yyyyyyyyyyy> where the number of C<y>'s are arbitrary, and are interpreted as a floating point value.

That is, C<1.001> is B<NOT> the same as C<1.01> and C<1.1>

However, Gentoo's version scheme sees C<1.001> similar to C<1.001.000> which is similar to C<1.1.0> and thus,
similar to C<1.1>.

Obviously this will not do, because when somebody says they need C<< >=1.05 (g:1.5) >> expecting C<< 1.06 (g:1.6) >>, but instead
get C<< 1.009 (g:1.9) >>, things will break.

Hence, detection of these cases and normalizing them is essential:

  1.001 -> 1.1.0
  1.01  -> 1.10.0
  1.1   -> 1.100.0
  1.05  -> 1.50.0
  1.06  -> 1.60.0
  1.009 -> 1.9.0

  1.9.0 < 1.50.0 < 1.60.0

The simplest use of this library is with the shipped tool, C<gentoo-perlmod-version.pl>

  gentoo-perlmod-version.pl --oneshot 1.06  # 1.6.0

=head1 FUNCTIONS

=head2 gentooize_version

    my $normalized = gentooize_version( $weird_version )

gentooize_version tries hard to mangle a version that is part of a CPAN dist into a normalized form
for Gentoo, which can be used as the version number of the C<ebuild>, while storing the original upstream version in the
C<ebuild>.

    CPAN: Foo-Bar-Baz 1.5
    print gentooize_version('1.5');  # -> 1.500.0
    -> dev-perl/Foo-Bar-Baz-1.500.0.ebuild
    cat dev-perl/Foo-Bar-Baz-1.500.0.ebuild
    # ...
    # MODULE_VERSION="1.5"
    # ...

Normal behavior accepts only sane non-testing versions, i.e.:

    0.1         -> 0.001.0
    0.001       -> 0.1.0
    1.1         -> 1.001.0
    1.123.13    -> 1.123.13

Etc.

This uses L<< C<version.pm>|version >> to read versions and to normalize them.

    0.1    # 0.100.0
    0.01   # 0.10.0
    0.001  # 0.1.0
    0.0001 # 0.0.100

So assuming Perl can handle your versions, they can be normalized.

=head3 lax level 1

    my $normalized = gentooize_version( $werid_version, { lax => 1 } );

B<EXPERIMENTAL:> This feature is still in flux, and the emitted versions may change.

This adds one layer of laxativity, and permits parsing and processing of "Developer Release" builds.

    1.10-TRIAL  # 1.100.0_rc
    1.11-TRIAL  # 1.110.0_rc
    1.1_1       # 1.110.0_rc

=head3 lax level 2

    my $normalized = gentooize_version( $werid_version, { lax => 2 } );

B<EXPERIMENTAL:> This feature is still in flux, and the emitted versions may change.

This adds another layer of laxativity, and permits parsing and processing of packages with versions not officially supported by
Perl.

This means versions such as

    1.6.A       # 1.6.10
    1.6.AA      # 1.6.370
    1.6.AAA      # 1.6.370.10
    1.6.AAAA      # 1.6.370.370

    1.6.A6FGHKE # 1.6.366.556.632.14

This is performed by some really nasty tricks, and treats the ASCII portion like a set of pairs.

    1.6.A6.FG.HK.E

And each ASCII pair is treated like a Base36 number.

    0 -> 0
    ....
    9 -> 9
    A -> 10
    ...
    Z -> 35

A6 is thus

    10 * 36 + 6 => 366

As you can see, its really nasty, and hopefully its not needed.

=head1 ENVIRONMENT

This module recognizes the environment variable GENTOO_PERLMOD_VERSION_OPTS for a few features.

These are mostly useful for system wide or user-wide policies that may be applicable for using this module, depending on where
it is used.

This field is split by white-space and each token has a meaning.

=head2 always_lax

  GENTOO_PERLMOD_VERSION_OPTS+=" always_lax=0 "
  GENTOO_PERLMOD_VERSION_OPTS+=" always_lax=1 "
  GENTOO_PERLMOD_VERSION_OPTS+=" always_lax=2 "
  GENTOO_PERLMOD_VERSION_OPTS+=" always_lax   "# same as always_lax=1
  GENTOO_PERLMOD_VERSION_OPTS+=" -always_lax  "# unset always_lax

This environment setting, if specified, overrides any specification of "lax" in the code. If this specified more than once,
the right-most one applies.

Specifying C<-always_lax> will unset the setting, making it behave as if it had not been previously specified.

=head2 taint_safe

  GENTOO_PERLMOD_VERSION_OPTS+=" taint_safe  " #on
  GENTOO_PERLMOD_VERSION_OPTS+=" -taint_safe " #off

As it stands, this module only emits messages via STDOUT/STDERR when an error occurs. For diagnosis, sometimes user provided
data can appear in this output.

Specifying this option will remove the information as specified by the user where possible, to eliminate this risk if this
is a security issue for you.

It is not a guarantee of safety, but merely a tool you might find useful, depending on circumstances.

=head2 carp_debug

  GENTOO_PERLMOD_VERSION_OPTS+=" carp_debug " #on
  GENTOO_PERLMOD_VERSION_OPTS+=" -carp_debug " #off

Lots of information is passed to our internal carp proxy that could aid in debugging a future problem.
To see this information instead of the simple message that is usually sent to C<Carp>, enable this option.

B<Note:> As values in the hashes that would be printed can come from users, C<carp_debug> is ignored if C<taint_safe> is on.

=head1 THANKS

=over 4

=item Torsten Veller - Inspiration for this Module and all the work on Gentoo Perl.

=item Vincent Pit - For solving most of the real bugs in this code before people tried to use them.

=back

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
